taichi/transforms/auto_diff.cpp

```diff
@@ -749,33 +970,42 @@ class MakeAdjoint : public ADTransform {
   // Accumulate [value] to the adjoint of [primal]
   void accumulate(Stmt *primal, Stmt *value) {
     auto alloca_ = adjoint(primal);
-    if (!alloca_ || alloca_->is<ConstStmt>())
+    if (!alloca_ || alloca_->is<ConstStmt>()) {
       return;  // primal may be int variable
+    }
     if (alloca_->is<AdStackAllocaStmt>()) {
       auto alloca = alloca_->cast<AdStackAllocaStmt>();
-      if (is_real(alloca->ret_type)) {
+      if (is_real(alloca->ret_type.get_element_type())) {
         insert<AdStackAccAdjointStmt>(alloca, load(value));
       }
     } else {
       TI_ASSERT(alloca_->is<AllocaStmt>());
       auto alloca = alloca_->as<AllocaStmt>();
       auto local_load = insert<LocalLoadStmt>(alloca);
+      local_load->ret_type = alloca->ret_type.ptr_removed();
       insert<LocalStoreStmt>(alloca, add(local_load, value));
     }
   }
 
   Stmt *adjoint(Stmt *stmt) {
-    if (!is_real(stmt->ret_type) || stmt->is<ConstStmt>()) {
+    DataType adjoint_dtype = stmt->ret_type.ptr_removed();
+    if (stmt->ret_type->is<TensorType>()) {
+      adjoint_dtype = TypeFactory::get_instance().get_tensor_type(
+          stmt->ret_type->as<TensorType>()->get_shape(), PrimitiveType::f32);
+    } else if (stmt->is<MatrixPtrStmt>()) {
```

Comment:
Why we skip the adjoint of `MatrixPtrStmt`?
e.g., in the case below, IIUC the adjoint of $val comes from that of $1, which is a adjoint of a `MatrixPtrStmt`
```
        $0 = Tensor alloca <4 x i32>
        $1 = matrix ptr $0, 2 // offset = 2
        $2 : local store $1, $val
```